pkg/controller/association/reconciler.go

```diff
@@ -330,10 +331,23 @@ func (r *Reconciler) reconcileAssociation(ctx context.Context, association commo
 	if err != nil {
 		return commonv1.AssociationPending, err
 	}
-	// Detect if we should use a service account. If it is the case create the related Secrets and update the association
-	// configuration on the associated resource.
-	assocLabels := r.AssociationResourceLabels(k8s.ExtractNamespacedName(association.Associated()), assocRef.NamespacedName())
+	// Detect if we should use a service account.
+	var esHints hints.OrchestrationsHints
 	if len(serviceAccount) > 0 {
+		// We must first ensure that the relevant orchestration hint is set on the Elasticsearch cluster.
+		esHints, err = hints.NewFrom(es)
+		if err != nil {
+			return commonv1.AssociationPending, err
+		}
+		if !esHints.ServiceAccounts.IsSet() {
+			r.log(k8s.ExtractNamespacedName(association)).Info("Waiting for Elasticsearch to report if service accounts are fully rolled out")
+			return commonv1.AssociationPending, nil
+		}
+	}
+
+	// If it is the case create the related Secrets and update the association configuration on the associated resource.
+	assocLabels := r.AssociationResourceLabels(k8s.ExtractNamespacedName(association.Associated()), assocRef.NamespacedName())
+	if len(serviceAccount) > 0 && esHints.ServiceAccounts.IsTrue() {
```

Comment:
Should this not be 
```diff
-       if len(serviceAccount) > 0 && esHints.ServiceAccounts.IsTrue() {
+       if len(serviceAccount) > 0 && esHints.ServiceAccounts.IsSet() {
                applicationSecretName := secretKey(association, r.ElasticsearchUserCreation.UserSecretSuffix)
                r.log(k8s.ExtractNamespacedName(association)).V(1).Info("Ensure service account exists", "sa", serviceAccount)
                err := ReconcileServiceAccounts(
@@ -367,8 +367,12 @@ func (r *Reconciler) reconcileAssociation(ctx context.Context, association commo
                expectedAssocConf.AuthSecretName = applicationSecretName.Name
                expectedAssocConf.AuthSecretKey = "token"
                expectedAssocConf.IsServiceAccount = true
-               // update the association configuration if necessary
-               return r.updateAssocConf(ctx, expectedAssocConf, association)
+
+               if esHints.ServiceAccounts.IsTrue() {
+                       // update the association configuration if necessary
+                       return r.updateAssocConf(ctx, expectedAssocConf, association)
+               }
+               return commonv1.AssociationPending, nil
        }
```
We can create the service token user immediately. There should be no need to check for the annotation at this point (but it does not hurt either). 

What needs to be predicated on the presence of the hint annotation is the actual update of the association config on the Kibana resource because that will trigger the rolling upgrade of the Kibana Pods. Updating the mounted secret before that with service account token information should be fine as Kibana does not currently hot reload its config (there is of course an edge case here if an unplanned restart of Kibana happens in while we are waiting for Elasticsearch to roll out the token)